### PR TITLE
pm emulation blog entry

### PR DESCRIPTION
**Pull request by Maciuch**

### CODE REVIEW

**Comment by krzycz** [comment]
Would it make sense to mention that in case of experimenting with NVML, you can also "emulate" persistent memory with ramdisk (i.e. */dev/shm*), or force pmem behavior by setting environment variable *PMEM_IS_PMEM_FORCE=1*.  This would eliminate the performance hit caused by *msync*.

**Comment by ykorman** [comment]
You should mention that it's not recommended to place the pmem range below the 4G mark - otherwise it might overlap with the i/o window (legacy for 32bit systems).
For example, if you define 'memmap=4G!3G' in a kvm-qemu VM, you'll get a weird setup:
```
root@ubuntu:~# dmesg
...
[    0.000000] user: [mem 0x0000000000100000-0x00000000bfffffff] usable
[    0.000000] user: [mem 0x00000000c0000000-0x00000000dfffdfff] persistent (type 12)
[    0.000000] user: [mem 0x00000000dfffe000-0x00000000dfffffff] persistent (type 12)
[    0.000000] user: [mem 0x00000000e0000000-0x00000000feffbfff] persistent (type 12)
[    0.000000] user: [mem 0x00000000feffc000-0x00000000feffffff] persistent (type 12)
[    0.000000] user: [mem 0x00000000ff000000-0x00000000fffbffff] persistent (type 12)
[    0.000000] user: [mem 0x00000000fffc0000-0x00000000ffffffff] persistent (type 12)
[    0.000000] user: [mem 0x0000000100000000-0x00000001bfffffff] persistent (type 12)
[    0.000000] user: [mem 0x00000001c0000000-0x000000021fffffff] usable
...

root@ubuntu:~# lsblk
NAME   MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
sr0     11:0    1  366K  0 rom
vda    253:0    0   20G  0 disk
└─vda1 253:1    0   20G  0 part /
pmem0m 259:0    0  512M  0 disk
pmem4m 259:1    0 15.8M  0 disk
pmem6m 259:2    0    3G  0 disk
```
(the qemu window is a bit less that 512M, for physical machines, it's usually ~2G)

Probably not what you wanted.

File: _posts/2016-02-22-pm-emulation.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+{% highlight C linenos %}
+#make nconfig
+	-> Device Drivers -> NVDIMM Support ->
+			<M>PMEM; <M>BLK; <*>BTT
+{% endhighlight %}
+
+You are ready to build your Kernel
+
+{% highlight C linenos %}
```
Comment (tomaszkapela) [inline _posts/2016-02-22-pm-emulation.md]: this looks more like `highlight sh`

Comment (Maciuch) [inline _posts/2016-02-22-pm-emulation.md]: fixing all to sh, and plain text, thanks

**Comment by Maciuch** [comment]
@krzycz Right, I should probably mention those options, however we don't want to elaborate much around those since they are not the preferred way to emulate PM.

@ykorman Thank for this, I shall add it!

File: _posts/2016-02-22-pm-emulation.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+#sudo vi /etc/default/grub
+GRUB_CMDLINE_LINUX="memmap=nn[KMG]!ss[KMG]"
+On BIOS-based machines:
+#sudo grub2-mkconfig -o /boot/grub2/grub.cfg
+On UEFI-based machines:
+#sudo grub2-mkconfig -o /boot/efi/EFI/centos/grub.cfg
+{% endhighlight %}
+
+After machne reboot you should be able to see the emulated device as `/dev/pmem0`.
```
Comment (tomaszkapela) [inline _posts/2016-02-22-pm-emulation.md]: machne -> machine

File: _posts/2016-02-22-pm-emulation.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+You are ready to build your Kernel
+
+{% highlight sh %}
+$ make -jX
+	where X is the number of cores on the machine
+{% endhighlight %}
+
+Install the kernel
+
+{% highlight sh %}
```
Comment (krzycz) [inline _posts/2016-02-22-pm-emulation.md]: A style nit - this is actually not a shell script, but rather a console output. Syntax highlighter treats lines starting with "#" as comments, so it doesn't look well.

Comment (Maciuch) [inline _posts/2016-02-22-pm-emulation.md]: true, changed to 'console', it basically does no highlighting and looks normal, you can take a look at the result here:
http://maciuch.github.io/2016/02/22/pm-emulation.html

**Pull request merged**